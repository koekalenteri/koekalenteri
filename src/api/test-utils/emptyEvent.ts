import type { ConfirmedEvent } from '../../types'

export const emptyEvent: ConfirmedEvent = {
  classes: [{ class: 'ALO' }],
  contactInfo: {
    official: {
      name: 'Teemu Toimitsija',
      email: 'official@example.com',
      phone: '040-official',
    },
    secretary: {
      name: 'Siiri Sihteeri',
      email: 'secretary@example.com',
      phone: '040-secretary',
    },
  },
  cost: 123,
  costMember: 123,
  createdAt: new Date('1989-01-01'),
  createdBy: 'test',
  description: 'test',
  endDate: new Date('1990-01-01'),
  entries: 0,
  entryEndDate: new Date('1990-01-01'),
  entryStartDate: new Date('1990-01-01'),
  eventType: 'test',
  id: 'test',
  judges: [123],
  kcId: 123456,
  location: 'test',
  modifiedAt: new Date('1989-01-02'),
  modifiedBy: 'test',
  name: 'test',
  official: {
    id: '1',
    name: 'Teemu Toimitsija',
  },
  priority: ['member'],
  secretary: {
    id: '2',
    name: 'Siiri Sihteeri',
  },
  organizer: {
    id: '1',
    name: 'Suomen Noutajakoirajärjestö ry',
  },
  places: 10,
  startDate: new Date('1990-01-01'),
  state: 'confirmed' as const,
}
