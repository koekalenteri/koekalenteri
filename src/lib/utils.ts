import type {
  DeepPartial,
  EventState,
  JsonValue,
  PublicContactInfo,
  PublicDogEvent,
  RegistrationDate,
  RegistrationTime,
} from '../types'

import { eachDayOfInterval, subDays } from 'date-fns'
import { diff } from 'deep-object-diff'

import { zonedEndOfDay, zonedStartOfDay } from '../i18n/dates'

type EventVitals = Partial<Pick<PublicDogEvent, 'startDate' | 'endDate' | 'entryStartDate' | 'entryEndDate' | 'state'>>

export const isValidForEntry = (state?: EventState) => !['draft', 'tentative', 'cancelled'].includes(state ?? '')

export const isEventUpcoming = ({ startDate }: EventVitals, now = new Date()) => !!startDate && startDate > now

export const isEntryUpcoming = ({ entryStartDate, state }: EventVitals, now = new Date()) =>
  !!entryStartDate && entryStartDate > now && (isValidForEntry(state) || state === 'tentative')

export const isEntryOpen = ({ entryStartDate, entryEndDate, state }: EventVitals, now = new Date()) =>
  !!entryStartDate &&
  !!entryEndDate &&
  zonedStartOfDay(entryStartDate) <= zonedEndOfDay(now) &&
  zonedEndOfDay(entryEndDate) >= zonedEndOfDay(now) &&
  isValidForEntry(state)

export const isEntryClosing = (event: EventVitals, now = new Date()) =>
  !!event.entryEndDate &&
  isEntryOpen(event, now) &&
  subDays(event.entryEndDate, 7) <= zonedEndOfDay(now) &&
  isValidForEntry(event.state)

export const isEntryClosed = ({ startDate, entryEndDate }: EventVitals, now = new Date()) =>
  !!startDate && !!entryEndDate && zonedEndOfDay(entryEndDate) < now && zonedStartOfDay(startDate) > now

export const isEventOngoing = ({ startDate, endDate, state }: EventVitals, now = new Date()) =>
  !!startDate &&
  !!endDate &&
  zonedStartOfDay(startDate) <= zonedEndOfDay(now) &&
  zonedEndOfDay(endDate) >= zonedEndOfDay(now) &&
  isValidForEntry(state) &&
  state !== 'confirmed'

export const isEventOver = ({ endDate }: EventVitals, now = new Date()) => !!endDate && zonedEndOfDay(endDate) < now

export const eventDates = (event?: Pick<PublicDogEvent, 'classes' | 'startDate' | 'endDate'> | null) => {
  if (!event) return []
  return event.classes.length
    ? uniqueDate(event.classes.map((c) => c.date ?? event.startDate))
    : eachDayOfInterval({ start: event.startDate, end: event.endDate })
}

export const uniqueClasses = (event?: Pick<PublicDogEvent, 'classes'> | null) =>
  unique((event?.classes ?? []).map((c) => c.class))

export const placesForClass = (
  event: DeepPartial<Pick<PublicDogEvent, 'places' | 'classes'>> | undefined | null,
  cls: string
) => {
  if (!event) {
    return 0
  }

  return (
    (event.classes ?? []).filter((c) => c.class === cls).reduce((acc, cur) => acc + (Number(cur.places) || 0), 0) ||
    Number(event.places) ||
    0
  )
}

export const uniqueClassDates = (event: PublicDogEvent, cls: string) =>
  cls === event.eventType
    ? eventDates(event)
    : uniqueDate(event.classes.filter((c) => c.class === cls).map((c) => c.date ?? event.startDate))

export const registrationDates = (event: PublicDogEvent, times: RegistrationTime[], cls?: string | null) =>
  (cls ? uniqueClassDates(event, cls) : eventDates(event)).flatMap<RegistrationDate>((date) =>
    times.map((time) => ({ date, time }))
  )

export const unique = <T = string>(arr: T[]): T[] => arr.filter((c, i, a) => a.indexOf(c) === i)

export const uniqueFn = <T>(arr: T[], cmp: (a: T, b: T) => boolean): T[] =>
  arr.filter((c, i, a) => a.findIndex((f) => cmp(f, c)) === i)

export const uniqueDate = (arr: Date[]) => [...new Set<number>(arr.map((d) => d.valueOf()))].map((v) => new Date(v))

const DATE_RE = /^\d{4}-(?:0[1-9]|1[0-2])-(?:[0-2][1-9]|[1-3]0|3[01])$/
const TIME_RE = /^(?:[0-1]\d|2[0-3])(?::[0-6]\d)(?::[0-6]\d)?(?:\.\d{3})?(?:[+-][0-2]\d:[0-5]\d|Z)?$/

export const isDateString = (value: unknown): value is string => {
  if (typeof value !== 'string') {
    return false
  }
  const [date, time] = value.split('T')
  return DATE_RE.test(date) && TIME_RE.test(time)
}

function dateReviver(_key: string, value: JsonValue): JsonValue | Date {
  if (isDateString(value)) {
    const dateObj = new Date(value)
    if (!isNaN(+dateObj)) {
      return dateObj
    }
  }
  return value
}

export const parseJSON = (json: string, reviveDates: boolean = true) => {
  if (!json) return undefined

  const reviver = reviveDates ? dateReviver : undefined

  return JSON.parse(json, reviver)
}

export type AnyObject = Record<string, any>
export type EmptyObject = Record<string, never>
export type Entries<T> = {
  [K in keyof T]: [K, T[K]]
}[keyof T][]

export const isEmpty = (o: AnyObject): o is EmptyObject => Object.keys(o).length === 0

export const isObject = (o: unknown): o is AnyObject =>
  o !== null && Object.prototype.toString.call(o) === '[object Object]'

export const isEmptyObject = (o: unknown): o is EmptyObject => isObject(o) && isEmpty(o)

export const hasChanges = (a: object | undefined | null, b: object | undefined | null): boolean =>
  !isEmptyObject(diff(a ?? {}, b ?? {}))

export const clone = <T extends AnyObject>(a: T): T => ({ ...a })

export const merge = <T extends AnyObject>(a: T, b: DeepPartial<T>): T => {
  const result = isObject(a) ? clone(a) : ({} as T)
  if (!isObject(b)) {
    return result
  }
  for (const [key, value] of Object.entries(b) as Entries<T>) {
    if (isObject(value)) {
      const old = result[key]
      result[key] = isObject(old) ? merge(old, value) : value
    } else {
      result[key] = value
    }
  }
  return result
}

// This function merges a patch object into a base object, but only creates new objects for paths that actually change
export const patchMerge = <T extends AnyObject, P extends DeepPartial<T>>(base: T, patch: P): T & P => {
  if (!isObject(base) || !isObject(patch)) return patch as T & P

  let changed = false
  const result = { ...base } as T & P
  for (const key of Object.keys(patch)) {
    const merged = patchMerge(base[key], (patch as any)[key])
    if (merged !== base[key]) {
      ;(result as any)[key] = merged
      changed = true
    }
  }

  return changed ? result : (base as T & P)
}

export const printContactInfo = (info?: PublicContactInfo) =>
  [info?.name, info?.phone, info?.email].filter(Boolean).join(', ')
