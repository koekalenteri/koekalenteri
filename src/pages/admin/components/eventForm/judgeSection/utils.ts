import type { DeepPartial, EventClass, EventType, Judge, PublicJudge } from '../../../../../types'
import type { PartialEvent } from '../../EventForm'

import { isSameDay } from 'date-fns'

export type PartialPublicJudge = Partial<PublicJudge>
export type PartialPublicJudgeValue = PartialPublicJudge | PartialPublicJudge[]

export const filterJudges = (
  judges: Judge[],
  eventJudges: PublicJudge[],
  id: number | undefined,
  eventType?: EventType
) =>
  judges
    .filter((j) => !eventType?.official || j.eventTypes.includes(eventType.eventType))
    .filter((j) => j.id === id || !eventJudges.find((ej) => ej.id === j.id))

export const hasJudge = (c: DeepPartial<EventClass>, id?: number) =>
  Array.isArray(c.judge) ? c.judge.find((j) => j.id === id) : c.judge?.id === id

export const filterClassesByJudgeId = (classes?: EventClass[], id?: number) => classes?.filter((c) => hasJudge(c, id))

export const toArray = (j?: PartialPublicJudge): PartialPublicJudge[] => (j ? [j] : [])
export const makeArray = (j?: PartialPublicJudgeValue) => (Array.isArray(j) ? [...j] : toArray(j))
export const selectJudge = (j?: PartialPublicJudgeValue, judge?: PublicJudge): PartialPublicJudge[] => {
  const a = makeArray(j)
  if (judge) {
    const index = a.findIndex((cj) => cj.id === judge.id)
    if (index === -1) {
      a.push(judge)
    } else {
      a[index] = judge
    }
  }
  return a
}

export const removeJudge = (j?: PartialPublicJudgeValue, id?: number): PartialPublicJudge[] => {
  const a = makeArray(j)
  return a.filter((cj) => cj.id !== id)
}

export const updateJudge = (
  event: PartialEvent,
  id: number | undefined,
  judge: PublicJudge | undefined,
  values?: DeepPartial<EventClass>[]
) => {
  const isSelected = (c: DeepPartial<EventClass>) =>
    values?.find((v) => event && isSameDay(v.date ?? event.startDate, c.date ?? event.startDate) && v.class === c.class)
  return event.classes?.map((c) => ({
    ...c,
    judge: isSelected(c) ? selectJudge(c.judge, judge) : removeJudge(c.judge, id),
  }))
}
