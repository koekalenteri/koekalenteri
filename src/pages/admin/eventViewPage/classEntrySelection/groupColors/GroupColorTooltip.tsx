import type { ReactElement } from 'react'
import type { RegistrationDate } from '../../../../../types'

import { useMemo } from 'react'
import { useTranslation } from 'react-i18next'
import Tooltip from '@mui/material/Tooltip'

interface Props {
  readonly children: ReactElement
  readonly selected: RegistrationDate[]
}

const GroupColorTooltip = ({ selected, children }: Props) => {
  const { t } = useTranslation()
  const title = useMemo(
    () =>
      selected.length &&
      'Sopivat ryhmÃ¤t: ' +
        selected
          .map(
            (s) => t('dateFormat.weekday', { date: s.date }) + (s.time ? ' ' + t(`registration.time.${s.time}`) : '')
          )
          .join(', '),
    [selected, t]
  )

  return <Tooltip title={title}>{children}</Tooltip>
}

export default GroupColorTooltip
