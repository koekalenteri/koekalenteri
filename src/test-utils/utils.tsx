import type { RenderOptions, RenderResult } from '@testing-library/react'
import type { Options } from '@testing-library/user-event/dist/types/options'
import type { UserEvent } from '@testing-library/user-event/dist/types/setup/setup'
import type { RouteObject, RouterInit } from 'react-router'
import type { RecoilValue } from 'recoil'

import { act, useEffect } from 'react'
import { createMemoryRouter, createRoutesFromElements, RouterProvider } from 'react-router'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import mediaQuery from 'css-mediaquery'
import { useRecoilValue } from 'recoil'

/**
 * Abstraction to avoid re-writing all tests for the time being
 * @see https://github.com/remix-run/react-router/blob/main/packages/react-router/__tests__/data-memory-router-test.tsx
 */
export function DataMemoryRouter({
  basename,
  children,
  hydrationData,
  initialEntries,
  initialIndex,
  routes,
}: {
  readonly basename?: RouterInit['basename']
  readonly children?: React.ReactNode | React.ReactNode[]
  readonly fallbackElement?: React.ReactNode
  readonly hydrationData?: RouterInit['hydrationData']
  readonly initialEntries?: string[]
  readonly initialIndex?: number
  readonly routes?: RouteObject[]
}) {
  const router = createMemoryRouter(routes ?? createRoutesFromElements(children), {
    basename,
    hydrationData,
    initialEntries,
    initialIndex,
  })
  return <RouterProvider router={router} />
}

export const flushPromises = async (timers: boolean = true) => {
  for (let i = 0; i <= 7; i++) {
    await act(async () => {
      if (timers) jest.runOnlyPendingTimers()
      await Promise.resolve()
    })
  }
}

export const createMatchMedia =
  (width: number) =>
  (query: string): MediaQueryList => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })

export function RecoilObserver<T>({
  node,
  onChange,
}: {
  readonly node: RecoilValue<T>
  readonly onChange: (value: T) => void
}) {
  const value = useRecoilValue(node)
  useEffect(() => onChange(value), [onChange, value])
  return null
}

export function renderWithUserEvents(
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  userEventOptions?: Options
): RenderResult & { user: UserEvent } {
  return {
    user: userEvent.setup(userEventOptions),
    ...render(ui, options),
  }
}
