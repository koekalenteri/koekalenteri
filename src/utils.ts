import type {
  DeepPartial,
  DogEvent,
  EventState,
  JsonValue,
  RegistrationDate,
  RegistrationTime,
  ShowContactInfo,
} from './types'

import { eachDayOfInterval, endOfDay, startOfDay, subDays } from 'date-fns'
import { diff } from 'deep-object-diff'
import { toASCII } from 'punycode'

type EventVitals = Partial<Pick<DogEvent, 'startDate' | 'endDate' | 'entryStartDate' | 'entryEndDate' | 'state'>>

export const isValidForEntry = (state?: EventState) => !['draft', 'tentative', 'cancelled'].includes(state ?? '')

export const isEventUpcoming = ({ startDate }: EventVitals, now = new Date()) => !!startDate && startDate > now

export const isEntryUpcoming = ({ entryStartDate, state }: EventVitals, now = new Date()) =>
  !!entryStartDate && entryStartDate > now && isValidForEntry(state)

export const isEntryOpen = ({ entryStartDate, entryEndDate, state }: EventVitals, now = new Date()) =>
  !!entryStartDate &&
  !!entryEndDate &&
  startOfDay(entryStartDate) <= now &&
  endOfDay(entryEndDate) >= now &&
  isValidForEntry(state)

export const isEntryClosing = (event: EventVitals, now = new Date()) =>
  !!event.entryEndDate &&
  isEntryOpen(event, now) &&
  subDays(event.entryEndDate, 7) <= endOfDay(now) &&
  isValidForEntry(event.state)

export const isEntryClosed = ({ startDate, entryEndDate }: EventVitals, now = new Date()) =>
  !!startDate && !!entryEndDate && endOfDay(entryEndDate) < now && startOfDay(startDate) > now

export const isEventOngoing = ({ startDate, endDate, state }: EventVitals, now = new Date()) =>
  !!startDate &&
  !!endDate &&
  startOfDay(startDate) <= now &&
  endOfDay(endDate) >= now &&
  isValidForEntry(state) &&
  state !== 'confirmed'

export const isEventOver = ({ endDate }: EventVitals, now = new Date()) => !!endDate && endOfDay(endDate) < now

export const eventDates = (event?: DogEvent) => {
  if (!event) return []
  return event.classes.length
    ? uniqueDate(event.classes.map((c) => c.date ?? event.startDate))
    : eachDayOfInterval({ start: event.startDate, end: event.endDate })
}

export const uniqueClasses = (event?: DogEvent) => unique((event?.classes ?? []).map((c) => c.class))

export const placesForClass = (event: DeepPartial<Pick<DogEvent, 'places' | 'classes'>> | undefined, cls: string) => {
  if (!event) {
    return 0
  }

  return (
    (event.classes ?? []).filter((c) => c.class === cls).reduce((acc, cur) => acc + (Number(cur.places) || 0), 0) ||
    Number(event.places) ||
    0
  )
}

export const uniqueClassDates = (event: DogEvent, cls: string) =>
  cls === event.eventType
    ? eventDates(event)
    : uniqueDate(event.classes.filter((c) => c.class === cls).map((c) => c.date ?? event.startDate))

export const registrationDates = (event: DogEvent, times: RegistrationTime[], cls?: string) =>
  (cls ? uniqueClassDates(event, cls) : eventDates(event)).flatMap<RegistrationDate>((date) =>
    times.map((time) => ({ date, time }))
  )

export const unique = <T = string>(arr: T[]): T[] => arr.filter((c, i, a) => a.indexOf(c) === i)

export const uniqueFn = <T>(arr: T[], cmp: (a: T, b: T) => boolean): T[] =>
  arr.filter((c, i, a) => a.findIndex((f) => cmp(f, c)) === i)

export const uniqueDate = (arr: Date[]) => [...new Set<number>(arr.map((d) => d.valueOf()))].map((v) => new Date(v))

const DATE_RE = /^\d{4}-(?:0[1-9]|1[0-2])-(?:[0-2][1-9]|[1-3]0|3[01])$/
const TIME_RE = /^(?:[0-1]\d|2[0-3])(?::[0-6]\d)(?::[0-6]\d)?(?:\.\d{3})?(?:[+-][0-2]\d:[0-5]\d|Z)?$/

export const isDateString = (value: unknown): value is string => {
  if (typeof value !== 'string') {
    return false
  }
  const [date, time] = value.split('T')
  return DATE_RE.test(date) && TIME_RE.test(time)
}

function dateReviver(_key: string, value: JsonValue): JsonValue | Date {
  if (isDateString(value)) {
    const dateObj = new Date(value)
    if (!isNaN(+dateObj)) {
      return dateObj
    }
  }
  return value
}

export const parseJSON = (json: string) => (json ? JSON.parse(json, dateReviver) : undefined)

export type AnyObject = Record<string, unknown>
export type EmptyObject = Record<string, never>
export type Entries<T> = {
  [K in keyof T]: [K, T[K]]
}[keyof T][]

export const isEmpty = (o: AnyObject): o is EmptyObject => Object.keys(o).length === 0

export const isObject = (o: unknown): o is AnyObject =>
  o !== null && Object.prototype.toString.call(o) === '[object Object]'

export const isEmptyObject = (o: unknown): o is EmptyObject => isObject(o) && isEmpty(o)

export const hasChanges = (a: object | undefined | null, b: object | undefined | null): boolean =>
  !isEmptyObject(diff(a ?? {}, b ?? {}))

export const clone = <T extends AnyObject>(a: T): T => ({ ...a })

export const merge = <T>(a: T, b: DeepPartial<T>): T => {
  const result = isObject(a) ? clone(a) : ({} as T)
  if (!isObject(b)) {
    return result
  }
  for (const [key, value] of Object.entries(b) as Entries<T>) {
    if (isObject(value)) {
      const old = result[key]
      // @ts-expect-error Argument of type 'T[keyof T] & AnyObject' is not assignable to parameter of type 'DeepPartial<T[keyof T] & AnyObject>'
      result[key] = isObject(old) ? merge(old, value) : value
    } else {
      result[key] = value
    }
  }
  return result
}

const USEREXP = /^[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+){0,4}$/i
const DOMAINEXP = /^(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.){1,4}[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$/i
export const validEmail = (email: string) => {
  const parts = email.split('@')

  if (parts.length !== 2) {
    return false
  }

  const [user, domain] = parts
  if (user.match(USEREXP) === null) {
    return false
  }

  const asciiDomain = toASCII(domain) // allow internationalized domain name
  if (asciiDomain.match(DOMAINEXP) === null) {
    return false
  }

  return true
}

export const printContactInfo = (info?: ShowContactInfo) =>
  [info?.name, info?.phone, info?.email].filter(Boolean).join(', ')
