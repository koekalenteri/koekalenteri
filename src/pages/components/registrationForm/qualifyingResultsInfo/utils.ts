import type { EventResultRequirement, EventResultRequirements, EventResultRequirementsByDate } from '../../../../rules'
import type { ManualTestResult, QualifyingResult } from '../../../../types'

import { nanoid } from 'nanoid'

import { unique } from '../../../../lib/utils'
import { objectContains } from '../validation'

const asArray = (v: EventResultRequirements | EventResultRequirement) => (Array.isArray(v) ? v : [v])

function findFirstMissing(requirements: EventResultRequirementsByDate | undefined, results: QualifyingResult[]) {
  if (!requirements || typeof requirements.rules === 'function') {
    return []
  }
  for (const rule of requirements.rules) {
    for (const opt of asArray(rule)) {
      const { count, ...rest } = opt
      if (results.filter((r) => objectContains(r, rest)).length < count) {
        return rest
      }
    }
  }
}

export function availableTypes(requirements?: EventResultRequirementsByDate, eventType?: string) {
  if (!requirements || typeof requirements.rules === 'function') {
    switch (eventType) {
      case 'NOME-B SM':
        return ['NOME-B']
      case 'NOME-A SM':
        return ['NOME-A', 'NOME-A KV']
      case 'NOWT SM':
        return ['NOWT']
    }
    return []
  }
  return unique(requirements.rules.flatMap((rule) => asArray(rule).map((opt) => opt.type)))
}

const availableSMResults = (type?: string, eventType?: string, hasKVA?: boolean) => {
  switch (eventType) {
    case 'NOME-B SM':
      return [hasKVA ? undefined : 'FI KVA-B', 'VOI1', 'VOI2', 'VOI3'].filter(Boolean)
    case 'NOWT SM':
      return [hasKVA ? undefined : 'FI KVA-WT', 'VOI1', 'VOI2', 'VOI3'].filter(Boolean)
    case 'NOME-A SM':
      if (type === 'NOME-A')
        return [hasKVA ? undefined : 'FI KVA-FT', 'A1 CERT', 'A1 RES-CERT', 'A1', 'A2', 'A3'].filter(Boolean)

      if (type === 'NOME-A KV') return ['EXC CACIT', 'EXC RES-CACIT', 'EXC', 'VG', 'G']
  }
  return []
}

export function availableResults(
  requirements?: EventResultRequirementsByDate,
  type?: string,
  eventType?: string,
  existingResults?: ManualTestResult[]
) {
  if (!requirements || typeof requirements.rules === 'function') {
    if (eventType?.includes(' SM')) {
      const hasKVA = existingResults?.some((r) => r.result.startsWith('FI KVA'))

      return availableSMResults(type, eventType, hasKVA)
    }
    return []
  }
  return unique(
    requirements.rules.flatMap((rule) =>
      asArray(rule)
        .filter((opt) => !type || opt.type === type)
        .map((opt) => (opt.cert ? 'CERT' : opt.result))
    )
  )
}

export function createMissingResult(
  requirements: EventResultRequirementsByDate | undefined,
  results: Array<QualifyingResult | ManualTestResult>,
  regNo: string
): ManualTestResult {
  const rule = findFirstMissing(requirements, results)
  return {
    id: nanoid(10),
    regNo,
    date: new Date(),
    official: false,
    qualifying: true,
    type: '',
    judge: '',
    location: '',
    result: '',
    class: '',
    ...rule,
  }
}

export function resultBorderColor(qualifying: boolean | undefined) {
  if (qualifying === true) {
    return 'success.light'
  }
  if (qualifying === false) {
    return 'error.main'
  }
}

export function getResultId(result: ManualTestResult | QualifyingResult) {
  if ('id' in result) {
    return result.id
  }
  return result.type + result.class + result.date.toISOString()
}
