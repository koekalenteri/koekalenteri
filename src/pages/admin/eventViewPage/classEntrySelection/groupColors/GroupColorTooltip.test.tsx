import { screen } from '@testing-library/react'

import { renderWithUserEvents } from '../../../../../test-utils/utils'

import GroupColorTooltip from './GroupColorTooltip'

describe('GroupColorTooltip', () => {
  const mockDate = new Date('2023-01-01')
  const mockSelected = [{ date: mockDate, time: 'ap' as const }]

  it('renders', async () => {
    const { container } = renderWithUserEvents(
      <GroupColorTooltip selected={mockSelected}>
        <div data-testid="child-element">Test Child</div>
      </GroupColorTooltip>
    )
    expect(container).toMatchSnapshot()
  })

  it('renders the child element', () => {
    renderWithUserEvents(
      <GroupColorTooltip selected={mockSelected}>
        <div data-testid="child-element">Test Child</div>
      </GroupColorTooltip>
    )

    expect(screen.getByTestId('child-element')).toBeInTheDocument()
    expect(screen.getByText('Test Child')).toBeInTheDocument()
  })

  it('creates tooltip with correct title for a single date/time', async () => {
    const { user } = renderWithUserEvents(
      <GroupColorTooltip selected={mockSelected}>
        <div data-testid="child-element">Test Child</div>
      </GroupColorTooltip>
    )

    // Hover over the child element to show the tooltip
    await user.hover(screen.getByTestId('child-element'))

    // Check that the tooltip is shown with the correct text
    expect(screen.getByLabelText('Sopivat ryhmät: eeeeee registration.time.ap')).toBeInTheDocument()
  })

  it('creates tooltip with correct title for multiple dates/times', async () => {
    const multipleSelected = [
      { date: mockDate, time: 'ap' as const },
      { date: mockDate, time: 'ip' as const },
    ]

    const { user } = renderWithUserEvents(
      <GroupColorTooltip selected={multipleSelected}>
        <div data-testid="child-element">Test Child</div>
      </GroupColorTooltip>
    )

    // Hover over the child element to show the tooltip
    await user.hover(screen.getByTestId('child-element'))

    // Check that the tooltip is shown with the correct text
    expect(
      screen.getByLabelText('Sopivat ryhmät: eeeeee registration.time.ap, eeeeee registration.time.ip')
    ).toBeInTheDocument()
  })

  it('handles empty selected array', () => {
    renderWithUserEvents(
      <GroupColorTooltip selected={[]}>
        <div data-testid="child-element">Test Child</div>
      </GroupColorTooltip>
    )

    // The component should still render the child
    expect(screen.getByTestId('child-element')).toBeInTheDocument()
  })
})
