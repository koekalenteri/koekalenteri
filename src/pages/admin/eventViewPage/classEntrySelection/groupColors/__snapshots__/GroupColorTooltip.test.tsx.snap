// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupColorTooltip renders 1`] = `
<div>
  <div
    aria-label="Sopivat ryhmÃ¤t: eeeeee registration.time.ap"
    class=""
    data-mui-internal-clone-element="true"
    data-testid="child-element"
  >
    Test Child
  </div>
</div>
`;
